### PR TITLE
CVE-2007-4559 Patch

### PR DESCRIPTION
**Pull request by TrellixVulnTeam**
# Patching CVE-2007-4559

Hi, we are security researchers from the Advanced Research Center at [Trellix](https://www.trellix.com). We have began a campaign to patch a widespread bug named CVE-2007-4559. CVE-2007-4559 is a 15 year old bug in the Python tarfile package. By using extract() or extractall() on a tarfile object without sanitizing input, a maliciously crafted .tar file could perform a directory path traversal attack. We found at least one unsantized extractall() in your codebase and are providing a patch for you via pull request. The patch essentially checks to see if all tarfile members will be extracted safely and throws an exception otherwise. We encourage you to use this patch or your own solution to secure against CVE-2007-4559. Further technical information about the vulnerability can be found in this [blog](https://www.trellix.com/en-us/about/newsroom/stories/research/tarfile-exploiting-the-world.html).

If you have further questions you may contact us through this projects lead researcher [Kasimir Schulz](mailto:kasimir.schulz@trellix.com).

### CODE REVIEW

**Comment by bdice** [comment]
I don’t have time to do a full review right now, but thank you for this! I am commenting to let the rest of the maintainer team know that this is a vulnerability I’ve heard about prior to this PR, and I think it’s worth patching in some way. I might like to refactor some portion of the fix but in principle I support it.

**Comment by tommy-waltmann** [comment]
A proper fix for this security bug should be in python's tarfile module, but until that happens, I'm in favor of this as a temporary security fix. Have the folks @TrellixVulnTeam started working towards getting this fixed upstream?

File: signac/contrib/utility.py [hunk 1]
```diff
@@ -366,7 +366,26 @@ def _extract(filename):
...
+                    prefix = os.path.commonprefix([abs_directory, abs_target])
+
+                    return prefix == abs_directory
+
+                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+
+                    for member in tar.getmembers():
+                        member_path = os.path.join(path, member.name)
+                        if not is_within_directory(path, member_path):
+                            raise Exception("Attempted Path Traversal in Tar File")
```
Comment (bdice) [inline signac/contrib/utility.py:384]:
```suggestion
                            raise RuntimeError("Paths extracted from the tar file must be in the current directory. Path traversal is not permitted.")
```

**Review by bdice — APPROVED** [review]

File: signac/contrib/utility.py [hunk 2]
```diff
@@ -366,7 +366,29 @@ def _extract(filename):
...
+                    return prefix == abs_directory
+
+                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+
+                    for member in tar.getmembers():
+                        member_path = os.path.join(path, member.name)
+                        if not is_within_directory(path, member_path):
+                            # Path traversal is not permitted to protect against CVE-2007-4559.
+                            raise RuntimeError(
+                                "Paths extracted from the tar file must be in the current directory."
```
Comment (bdice) [inline signac/contrib/utility.py:386]:
```suggestion
                                "Paths extracted from the tar file are not permitted to traverse directories."
```

File: signac/contrib/utility.py [hunk 3]
```diff
@@ -366,7 +366,29 @@ def _extract(filename):
...
+                    return prefix == abs_directory
+
+                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+
+                    for member in tar.getmembers():
+                        member_path = os.path.join(path, member.name)
+                        if not is_within_directory(path, member_path):
+                            # Path traversal is not permitted to protect against CVE-2007-4559.
+                            raise RuntimeError(
+                                "Paths extracted from the tar file are not permitted to traverse directories."
```
Comment (bdice) [inline signac/contrib/utility.py:386]:
```suggestion
                                "Paths extracted from the tar file are not permitted to traverse "
                                "directories."
```

**Comment by bdice** [comment]
It appears that the best fix for this is to remove the function. `_extract` is never called in the codebase.

**Pull request merged**